src/main/java/com/example/nasapic/controller/PictureController.java

```diff
@@ -0,0 +1,39 @@
+package com.example.nasapic.controller;
+
+import com.example.nasapic.service.PictureService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/api")
+@RequiredArgsConstructor
+public class PictureController {
+    private final PictureService service;
+    @Value("${nasa.api.url}")
+    private String URL;
+    @Value("${nasa.api.key}")
+    private String KEY;
+
+    @GetMapping({"/mars/pictures/largest/{sol}","/mars/pictures/largest/{sol}/{camera}"})
+    public ResponseEntity<byte[]> getLargestPicture(@PathVariable String sol, @PathVariable(required = false) String camera) {
+        var url = UriComponentsBuilder.fromHttpUrl(URL)
```

Comment:
I would move this logic to the service layer. In this case, the controller should not know about the implementation details of third-party APIs, since PictureService is fully responsible for the integration with NasaApi, it should incapsulate all the logic of this integration within itself.

Reply:
Hey, thanks for feedback. I changed my code check it out another one please! I also create similar implementation using web client and non blocking reactive stack.

Reply:
@KirillKotenok good job👍 Just one small note about naming. The entity is named **Photo**, but the Service is a **Picture**Service. It can get confusing in some places. For example, the createPhoto() method, or NoSuchElementException, where we are working with photos in the stream, but the exception telling us that we can`t find **pictures***.
But this is not directly related to the topic of the lesson. Just a small tip.😉